share/eds/p2p/client.go

```diff
@@ -0,0 +1,143 @@
+package p2p
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+
+	"github.com/celestiaorg/celestia-node/share"
+	"github.com/celestiaorg/celestia-node/share/eds"
+	p2p_pb "github.com/celestiaorg/celestia-node/share/eds/p2p/pb"
+
+	"github.com/celestiaorg/go-libp2p-messenger/serde"
+	"github.com/celestiaorg/rsmt2d"
+)
+
+var log = logging.Logger("shrex/eds")
+
+var errNoMorePeers = errors.New("all peers returned invalid responses")
+
+// Client is responsible for requesting EDSs for blocksync over the ShrEx/EDS protocol.
+// This client is run by light nodes and full nodes. For more information, see ADR #13
+type Client struct {
+	protocolID protocol.ID
+	host       host.Host
+}
+
+// NewClient creates a new ShrEx/EDS client.
+func NewClient(host host.Host, protocolSuffix string) *Client {
+	return &Client{
+		host:       host,
+		protocolID: protocolID(protocolSuffix),
+	}
+}
+
+// RequestEDS requests the full EDS from one of the given peers.
+//
+// The peers are requested in a round-robin manner with retries until one of them gives a valid
+// response. Blocks forever until the context is canceled or a valid response is given.
+func (c *Client) RequestEDS(
+	ctx context.Context,
+	dataHash share.DataHash,
+	peers peer.IDSlice,
+) (*rsmt2d.ExtendedDataSquare, error) {
+	req := &p2p_pb.EDSRequest{Hash: dataHash}
+
+	// requests are retried for every peer until a valid response is received
+	excludedPeers := make(map[peer.ID]struct{})
+	for {
+		// if no peers are left, return
+		if len(peers) == len(excludedPeers) {
+			return nil, errNoMorePeers
+		}
+
+		for _, to := range peers {
+			// skip over excluded peers
+			if _, ok := excludedPeers[to]; ok {
+				continue
+			}
+
+			eds, err := c.doRequest(ctx, req, dataHash, to)
+			if eds != nil {
+				return eds, nil
+			}
+			if errors.Is(err, context.DeadlineExceeded) || errors.Is(err, context.Canceled) {
+				return nil, ctx.Err()
+			}
```

Comment:
Also cc @walldiss, for the stream helper you are working on.

Reply:
This error does happen, and is covered in two tests

Reply:
I see. When I wrote this I thought specifically about Read and Write which do not return context errors. The error is returned by either NewStream or ReadEDS inside doRequest. 

The problem with current code is that it does not cover for errors coming once deadline is reached on blocking Read or Write. The error returned there is net.Error and we should swap it out with context error and return. Otherwise, we go on the next try even when the operation is cancelled.

Reply:
@Wondertan, if you meant [this one ](https://github.com/celestiaorg/celestia-node/blob/3a58679ed84da966d01173f32780134c7b830594/header/p2p/session.go#L113), then it's ok and we should not return err there.

Reply:
No, not this one @vgonkivs

Reply:
Yeah, check how @distractedm1nd made it work https://github.com/celestiaorg/celestia-node/pull/1431/files#diff-677f2124709be3e4dbc7f0df73e8e4914f4d08cc80869fec9c55a74071fcd7d6R77-R80

Reply:
so, [this approach](https://github.com/celestiaorg/celestia-node/pull/1431/files#diff-677f2124709be3e4dbc7f0df73e8e4914f4d08cc80869fec9c55a74071fcd7d6R77-R80) is basically to handle mocknet errors?

Reply:
Not only, also for the yamux and quic errors in production when context is canceled by a deadline

Reply:
Ok, will add the same there. Thanks